xls/modules/fpsqrt_32.x

```diff
@@ -0,0 +1,232 @@
+//                             i+1   i             i+1   i
+//
+//      With some algebraic manipulation, it is not difficult to see
+//      that (2) is equivalent to
+//                             -(i+1)
+//                      s  +  2       <= y                       (3)
+//                       i                i
+//
+//      The advantage of (3) is that s  and y  can be computed by
+//                                    i      i
+//      the following recurrence formula:
+//          if (3) is false
+//
+//          s     =  s  ,       y    = y   ;                     (4)
+//           i+1      i          i+1    i
+//
+//      otherwise,
+//                         -i                      -(i+1)
+//          s     =  s  + 2  ,  y    = y  -  s  - 2              (5)
+//           i+1      i          i+1    i     i
+//
+//      One may easily use induction to prove (4) and (5).
+//      Note. Since the left hand side of (3) contain only i+2 bits,
+//            it is not necessary to do a full (53-bit) comparison
+//            in (3).
+//   3. Final rounding
+//      After generating the 53 bits result, we compute one more bit.
+//      Together with the remainder, we can decide whether the
+//      result is exact, bigger than 1/2ulp, or less than 1/2ulp
+//      (it will never equal to 1/2ulp).
+//      The rounding mode can be detected by checking whether
+//      huge + tiny is equal to huge, and whether huge - tiny is
+//      equal to huge for some floating point number "huge" and "tiny".
+//
+// Notes:  Rounding mode detection omitted.
+
+// This file implements [most of] IEEE 754 single-precision
+// floating point square-root, with the following exceptions:
+//  - Input denormals are treated as/flushed to 0.
+//      (denormals-are-zero / DAZ).
+//  - Only round-to-nearest mode is supported.
+//  - No exception flags are raised/reported.
+// In all other cases, results should be identical to other
+// conforming implementations, modulo exact significand
+// values in the NaN case (we emit a single, canonical 
+// representation for NaN (qnan) but accept all NaN 
+// respresentations as input).
+
+import float32
+import xls.dslx.stdlib.apfloat
+
+type APFloat = apfloat::APFloat;
+type F32 = float32::F32;
+
+// Determines if the given value is 0, taking into account
+// flushing subnormals.
+fn is_zero<EXP_SZ: u32, SFD_SZ: u32>(x: APFloat<EXP_SZ, SFD_SZ>) -> u1 {
+  x.bexp == uN[EXP_SZ]:0
+}
+
+pub fn fpsqrt_32(x: F32) -> F32 {
+  // Flush subnormal input.
+  let x = float32::subnormals_to_zero(x);
+
+  let exp = float32::unbiased_exponent(x);
+
+  let ix = u1:0 ++ u8:1 ++ x.sfd;
+  // If odd exp, double x to make it even.
+  let ix = ix << u32:1 if exp[0:1] else ix;
+  // exp = exp / 2, exponent of square root
+  let exp = exp >> u8:1;
+
+  // Generate sqrt(x) bit by bit.
+  let ix = ix << u32:1;
+  let (ix, r, q, s) = for (idx, (ix, r, q, s)): (u32, (u32, u32, u32, u32)) in range(u32:0, u32:23 + u32:2) {
```

Comment:
Maybe line-wrap, I guess. IDK  how to make this more readable...put ix, r, q, and s in a struct, maybe? (with more-symbolic names?). 

You don't have to do any of these; I'm not sure what could help, really. Consider if you can make it more pleasant to read :)

Reply:
Wrapped!
I started bundling all the loop carries in a single struct, bit it seemed like this was creating more clutter than it eliminated (struct def + construct initial value + struct.thing for reads in loop + re-pack into new struct at end of loop + pull out a couple fields after the loop)